### PR TITLE
Refactor RNNs

### PR DESCRIPTION
**Pull request by unnonouno**
I refactored RNNs. All NStepRNNs have similar codes. I extracted common codes from these implementation.
This fix is required before fixing interface of RNNs because we need to fix all interfaces of RNNs.

### CODE REVIEW

File: chainer/functions/connection/n_step_gru.py [hunk 1]
```diff
@@ -313,70 +309,23 @@ def n_step_gru_base(n_layers, dropout_ratio, hx, ws, bs, xs,
...
-                           six.moves.zip(h_forward, h_backward)]
-                hy.append(h)
-            else:
-                # Uni-directional GRU
-                xs_next = h_forward
-
-        ys = xs_next
-        hy = stack.stack(hy)
-        return hy, tuple(ys)
+        def f(x, h, c, w, b):
```
Comment (okuta) [inline chainer/functions/connection/n_step_gru.py]: Can you define this function in module level?

File: chainer/functions/connection/n_step_lstm.py [hunk 2]
```diff
@@ -445,80 +445,14 @@ def n_step_lstm_base(
...
-                           zip(h_forward, h_backward)]
-            else:
-                # Uni-directional RNN
-                xs_next = h_forward
-
-        ys = xs_next
-        hy = stack.stack(hy)
-        cy = stack.stack(cy)
-        return hy, cy, tuple(ys)
+        def f(x, h, c, w, b):
```
Comment (okuta) [inline chainer/functions/connection/n_step_lstm.py]: ditto.

File: chainer/functions/connection/n_step_rnn.py [hunk 3]
```diff
@@ -825,68 +824,93 @@ def n_step_rnn_base(n_layers, dropout_ratio, hx, ws, bs, xs,
...
+                xs = xs_next
             else:
-                # Uni-directional RNN
-                xs_next = h_forward
+                xs = _dropout_sequence(xs_next, dropout_ratio)
+            h, c, h_backward = _one_directional_loop(
+                f, reversed(xs), hx[idx], cx[idx], ws[idx], bs[idx])
+            h_backward.reverse()
+            # Concat
+            xs_next = [concat.concat([hfi, hbi], axis=1) for (hfi, hbi) in
```
Comment (okuta) [inline chainer/functions/connection/n_step_rnn.py]: Please remove parenthesis.

File: chainer/functions/connection/n_step_rnn.py [hunk 4]
```diff
@@ -825,68 +824,93 @@ def n_step_rnn_base(n_layers, dropout_ratio, hx, ws, bs, xs,
...
+        cy = stack.stack(cy)
+    else:
+        cy = None
+    return hy, cy, tuple(ys)
+
+
+def _one_directional_loop(f, xs, h, c, w, b):
+    h_list = []
+    for x in xs:
+        batch = x.shape[0]
```
Comment (okuta) [inline chainer/functions/connection/n_step_rnn.py]: How about using `len(x)`?

File: chainer/functions/connection/n_step_rnn.py [hunk 5]
```diff
@@ -825,68 +824,93 @@ def n_step_rnn_base(n_layers, dropout_ratio, hx, ws, bs, xs,
...
+        need_split = h.shape[0] > batch
```
Comment (okuta) [inline chainer/functions/connection/n_step_rnn.py]: How about using `len(h)`?

**Comment by unnonouno** [comment]
fixed

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
LGTM!

**Pull request merged**